code/addons/interactions/src/preview.test.ts

```diff
@@ -0,0 +1,34 @@
+import { expect, test } from 'vitest';
+import { fn, isMockFunction } from '@storybook/test';
+import { action } from '@storybook/addon-actions';
+
+import { traverseArgs } from './preview';
+
+test('traverseArgs', () => {
+  const args = {
+    deep: {
+      deeper: {
+        fnKey: fn(),
+        actionKey: action('name'),
+      },
+    },
+    arg2: Object.freeze({ frozen: true }),
+  };
+
+  expect(args.deep.deeper.fnKey.getMockName()).toEqual('spy');
+
+  const traversed = traverseArgs(args) as typeof args;
+  expect(traversed).toEqual({
+    deep: {
+      deeper: {
+        fnKey: args.deep.deeper.fnKey,
+        actionKey: args.deep.deeper.actionKey,
+      },
+    },
+    arg2: args.arg2,
+  });
+
+  expect(traversed.deep.deeper.fnKey.getMockName()).toEqual('fnKey');
```

Comment:
I'm not entirely sure how this test

1. ensures that args are mutated
2. ensures readonly args are _not_ mutated

Maybe 2 is just by seeing we don't get an error?

Maybe you could just add some clarifying comments. this test looks like something that in 6 month you have no idea why it's written this way.